enable 2 erisc mode for TT-Fabric on Blackhole (behind env var)

SeanNijjar opened the pull request:
This PR includes the TT-Fabric router changes necessarily to functionally support 2-Erisc parallelism of TT-Fabric router on Blackhole. The changes are tested lightly and functional, but are dependent on dispatcher changes in https://github.com/tenstorrent/tt-metal/pull/23301. Therefore, this PR pushes the updated code-paths disabled by default.  The user can opt-in to enable 2-Erisc execution of fabric with the environment variable `TT_METAL_FABRIC_BLACKHOLE_TWO_ERISC=1`.

To get an understanding of the "out-of-the-box" performance going from 1-erisc to 2-erisc, we jump from ~17 GB/s/dir/link @4k packet size on the neighbour exchange pattern to ~24 GB/s/dir/link. Performance will continue to improve from future work,.

There are several limitations to be aware of with the current implementation:
- Dynamic noc mode isn't supported yet on Ethernet cores
  - Therefore, each erisc is strictly limited to using exactly only one noc for the time being
  - Even if two eriscs want to use different noc command buffers, it's not possible if it's on the same noc
- All data-structures are stored in erisc local memory
  - Therefore for the time being, the only supported parallelization strategy is for one risc erisc to implement all sender channels and for the other to implement all receiver channels
    - It's not possible to implement sender channels on different eriscs.

Overall, this is the very first set of changes and many more changes will follow in the future.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI: https://github.com/tenstorrent/tt-metal/actions/runs/16335307547
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI: https://github.com/tenstorrent/tt-metal/actions/runs/16378533799
- [x] Blackhole LLMBox:https://github.com/tenstorrent/tt-metal/actions/runs/16335314726
  - same failure on main 
- [x] Blackhole nightly: https://github.com/tenstorrent/tt-metal/actions/runs/16354851003
- [x] T3K unit, frequent, demo, nightly: https://github.com/tenstorrent/tt-metal/actions/runs/16335328176
- [x] Microbenchmark: https://github.com/tenstorrent/tt-metal/actions/runs/16335330618

SeanNijjar commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -50,25 +50,97 @@ namespace tt::tt_fabric {
+            if (risc_id == 0) {
+                // ERISC0: Handle sender channels only
+                enable_handshake = true;
+                enable_context_switch = false;
```
So in this PR, should I add a `call_base_fw` option here and pass it down and then in the fabric main loop call this `service_base_fw` in place of context switch when this flag is true?

SeanNijjar replied:
Since we're not removing any functionality related to this topic, can I instead open an issue to address this in a separate PR?

SeanNijjar replied:
https://github.com/tenstorrent/tt-metal/issues/25345

SeanNijjar commented on tt_metal/fabric/hw/inc/edm_fabric/1d_fabric_constants.hpp:
```diff
@@ -302,7 +310,11 @@ constexpr size_t DEFAULT_NUM_ETH_TXQ_DATA_PACKET_ACCEPT_AHEAD = get_compile_time
-constexpr size_t SPECIAL_MARKER_0_IDX = MAIN_CT_ARGS_IDX_5 + 14;
+constexpr size_t MY_ERISC_ID = get_compile_time_arg_val(MAIN_CT_ARGS_IDX_5 + 14);
+constexpr size_t NUM_ACTIVE_ERISCS = get_compile_time_arg_val(MAIN_CT_ARGS_IDX_5 + 15);
+static_assert(MY_ERISC_ID < NUM_ACTIVE_ERISCS, "MY_ERISC_ID must be less than NUM_ACTIVE_ERISCS");
```
Sounds good. We may end up keeping these IDs so we can treat the IDs as logical rather than to specific eriscs if we decide to tie functionality to logical eriscs rather than the physical erisc0/1 if that makes sense. For example, maybe logical core0 handles sender channels but we could map that to either erisc0 or erisc1. TBD but I'll keep this in mind.

SeanNijjar commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1787,9 +1805,74 @@ void populate_local_sender_channel_free_slots_stream_id_ordered_map(
+        }
+    }
+    // TODO: ADD TEARDOWN SYNCHRONIZATION
+    if constexpr (IS_TEARDOWN_MASTER()) {
```
good catch thanks.

yugaoTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1787,9 +1805,71 @@ void populate_local_sender_channel_free_slots_stream_id_ordered_map(
+        RECEIVER_NUM_BUFFERS_ARRAY[NUM_RECEIVER_CHANNELS - 1],
+        NUM_TRANSACTION_IDS,
+        NUM_TRANSACTION_IDS> receiver_channel_1_trid_tracker) {
+    if constexpr (NUM_ACTIVE_ERISCS > 1) {
+        wait_for_other_local_erisc();
+    }
+    if constexpr (is_receiver_channel_serviced[0]) {
+        receiver_channel_0_trid_tracker.all_buffer_slot_transactions_acked();
+    }
+    if constexpr (is_receiver_channel_serviced[1]) {
+        receiver_channel_1_trid_tracker.all_buffer_slot_transactions_acked();
+    }
```
should we move wait_for_other_local_erisc after the receiver barrier on the ack?
in that way we can make sure before we reinit the local noc counters, we already have all the responses back.

SeanNijjar replied:
I will add `static_assert(noc_mode != DM_DYNAMIC_NOC, "The fabric router implementation doesn't support dynamic noc mode. The implementation must be updated to support this")` to address this concern

SeanNijjar replied:
on second thought, to keep the review lean, I will just make one of the risc cores do the noc counters init (otherwise I need to change `noc_nonblocking_api.h` and kickoff another full test cycle (WH + BH))

yugaoTT replied:
so I guess a question is when we reset the local NoC counters, is it safe for sender channels to have wrong counter values?

for example, a sequence like this:
wait_for_other_local_erisc
receiver wait for ACKs to be back
before the ACK is back, sender core reset local counter (NoC1 will give a smaller value since ack is not back yet)
receiver core reset local counter

SeanNijjar replied:
the receiver wouldn't receive the ack for the other noc since it only ever uses `noc_index`. This is consistent with current behaviour where we have essentially garbage data in our local noc counters for the noc that isn't our `noc_index`. In that case we'd just be reading the current values for that other noc, and storing them locally, but really those values never get used (unless in dynamic noc mode)

yugaoTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -2180,18 +2279,21 @@ void kernel_main() {
-                    .template setup_edm_noc_cmd_buf<
-                        tt::tt_fabric::edm_to_downstream_noc,
-                        tt::tt_fabric::forward_and_local_write_noc_vc>();
+                // Only receiver channel servicing cores should be setting up the noc cmd buf.
+                if constexpr (NUM_ACTIVE_ERISCS == 1) {
```
a bit confused about the comment and actual condition, should it be is_receiver_channel_serviced?

yugaoTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -2240,10 +2342,14 @@ void kernel_main() {
-                    tt::tt_fabric::edm_to_downstream_noc,
-                    tt::tt_fabric::forward_and_local_write_noc_vc>();
+
+            // Only receiver channel servicing cores should be setting up the noc cmd buf.
+            if constexpr (NUM_ACTIVE_ERISCS == 1) {
+                downstream_edm_noc_interfaces[NUM_USED_RECEIVER_CHANNELS - 1]
```
same as above

SeanNijjar replied:
Added a comment to clarify.
```
            // Only receiver channel servicing cores should be setting up the noc cmd buf.
            // If there is only one active erisc, then it is guaranteed we are the receiver channel
            // servicing core. Otherwise, in multi-erisc mode, this initialization happens later due
            // to a noc dependency. See the comment for the initialization that happens later in multi-erisc mode.
```            

There is a comment further down that explains the dependency:
```
        // Two erisc mode requires us to reorder the cmd buf programming/state setting
        // because we need to reshuffle some of our cmd_buf/noc assignments around for
        // just the fabric bringup phase. These calls are also located earlier for the
        // single erisc mode"
```

yugaoTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -2279,17 +2387,31 @@ void kernel_main() {
-            ASSERT(
-                get_ptr_val(downstream_edm_noc_interfaces[i].worker_credits_stream_id) ==
-                DOWNSTREAM_SENDER_NUM_BUFFERS);
+        if constexpr (is_receiver_channel_serviced[0]) {
```
should it be is_receiver_channel_serviced[0] || is_receiver_channel_serviced[1] ? if we already decided receiver channels won't be split across cores then it doesn't matter.

yugaoTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -2279,17 +2387,31 @@ void kernel_main() {
         }
     }
 
+    if constexpr (NUM_ACTIVE_ERISCS > 1) {
+        wait_for_other_local_erisc();
+    }
```
could you please add a comment here why a sync is required?

yugaoTT reviewed (approved):
lgtm

SeanNijjar committed:
add some comments for special behaviour

SeanNijjar committed:
add BH 1d bw tests

fix to avoid remote device usage in test (due to ctx switch)

remove bad character

get line devices

add initial BH perf tests

Only 2-chip tests are added until 4-chip tests can be validated on 4chip machines.

Note when reading bandwidth numbers for BH, that the bandwidth is reported as B/c. Therefore, when understanding the real bandwidth you must account for clock frequency (e.g. multiply by 1.35 on BH to see that effective bandwidth is actually 17.2 GB/s/dir for the 4K size test case).

sq

sq

sq

enable 2-erisc fabric router on BH

fixes

Debug (DON'T MERGE)

BH fabric use second txq for 2-erisc mode and add 2-erisc safe teardown sequence

fix fabric router intiailization bugs when two eriscs are enabled

support other machine configs, like those found w/ BH.

fix

fabric use  2erisc on BH fixes

> Conflicts:
>	.github/workflows/blackhole-llmbox-unit-tests-impl.yaml
>	tt_metal/fabric/erisc_datamover_builder.cpp
>	tt_metal/fabric/hw/inc/edm_fabric/1d_fabric_constants.hpp

> Conflicts:
>	tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp

DON'T MERGE (debug)

assign configured noc differently for each erisc

some candidate fixes

don't run 2-erisc mode by default. Instead enable with environment variable

specify TT_METAL_FABRIC_BLACKHOLE_TWO_ERISC=1 to enable 2-erisc mode in TT-fabric (Blackhole). This support is still experimental.

cleanup

disable 2-erisc test because support isn't mainlined yet

how did it build before... cache?

revert accidental change

corrections to benchmark script

cleanup

cleanup

SeanNijjar committed:
revert accidental change

SeanNijjar committed:
address feedback

SeanNijjar committed:
resolve performance regression

SeanNijjar commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1657,15 +1657,33 @@ template <typename EdmChannelWorkerIFs>
+                }
+            });
+    } else {
+        // Very slight performance regression on WH if we commonize to the above path, so we preserve this path
+        // too
```
I guess we'll find out when I try

SeanNijjar committed:
fix misconfigured attributed

SeanNijjar committed:
cleanup
